### PR TITLE
Performance Testing Implementation

### PR DESCRIPTION
**Pull request by dkosteck**
Notes: 
- This relies on https://github.com/redhat-eets/netgauge/pull/9 being merged before it will work
- trafficgen PCI addresses are required in testbed.yaml
- additional fields are required for config.yaml (see readme or template)

### CODE REVIEW

**Comment by dkosteck** [comment]
Report: https://gist.github.com/dkosteck/2aefed7c8cd38fb9670be8936a5b28d3

**Review by akiselev1 — CHANGES_REQUESTED** [review]
Can we run e2e-test CI once https://github.com/redhat-eets/netgauge/pull/9 is merged?

File: sriov/tests/SR_IOV_Performance/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+```
+echo 1 > /sys/class/net/$PF/device/sriov_numvfs
+ip link set $PF vf 0 spoof off
+ip link set $PF vf 0 trust on
+echo $VF_PCI > /sys/bus/pci/devices/$VF_PCI/driver/unbind
+echo vfio-pci > /sys/bus/pci/devices/$VF_PCI/driver_override
+echo $VF_PCI > /sys/bus/pci/drivers/vfio-pci/bind
+```
+
+* On the trafficgen server, start the prebuilt testpmd container
```
[the remark this answers is not in the document]
Comment (jianzzha) [reply sriov/tests/SR_IOV_Performance/README.md:33]: we have been using the DUT in all other test case document

File: README.md [hunk 2]
```diff
@@ -67,19 +67,24 @@ dut:
   interface:
     pf1:
       name: "ens7f0"        # first PF interface name
+      pci: "0000:00:00.0"   # first PF PCI address
```
Comment (jianzzha) [inline README.md:70]: we used to have the "pci" info in the yaml file, then we decided we could derive this info from the interface name; shall we use that util function to get this pci info?

File: sriov/tests/config_template.yaml [hunk 3]
```diff
@@ -11,3 +11,10 @@ container_volumes: "-v /sys:/sys -v /dev:/dev -v /lib/modules:/lib/modules"
 vlan: 10
 mtu: 1500
 bonding_switch_delay: 5
+testpmd_img: "testpmd:22.11.2"
+testpmd_port: 9000
+trafficgen_img: "trafficgen:latest"
+trafficgen_port: 8080
+trafficgen_ip: 10.88.0.88
```
Comment (jianzzha) [inline sriov/tests/config_template.yaml:18]: Normally we don't need to specify this - the trafficgen container will use its host ip address, the rely on the port mapping ( '-p' that is being used ). If the client is on the same host as the trafficgen container, then client simply points to the localhost as the server ip and the '-p' host port. If the client is from another machine, then the client will point to the trafficgen host ip as the server ip and the   '-p' host port.

**Review by jianzzha — COMMENTED** [review]
@dkosteck It looks fine except two places arguable:
1) whether 'trafficgen_ip" is required, as it really can be just the trafficgen host ip address
2) where "pci" is necessary, as that info can be derived from the interface name using the util function

**Comment by jianzzha** [comment]
@PatrickKutch

File: sriov/tests/SR_IOV_Performance/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkosteck) [reply sriov/tests/SR_IOV_Performance/README.md]: Updated in latest commit

File: README.md (hunk 2 above)
Comment (dkosteck) [reply README.md]: Addressed in the latest commit

**Comment by PatrickKutch** [comment]
why are we mounting /lib/firmware into the container?
-v /lib/firmware:/lib/firmware

**Comment by PatrickKutch** [comment]
client_cmd = [
        "wget -O /tmp/client.py "
        "https://raw.githubusercontent.com/redhat-eets/netgauge/main/rfc2544/client.py"
    ]
Will fail if we are doing tests in a air-gapped test environment (which we frequently do) - why is this not part of the container dockerfile?

**Comment by PatrickKutch** [comment]
client_cmd = (
        f"python3 /tmp/client.py status "
        f"--server-addr {settings.config['trafficgen_ip']} "
        f"--server-port {settings.config['trafficgen_port']}"
    )

Any concern over hard coding 'python3'?

**Comment by jianzzha** [comment]
> why are we mounting /lib/firmware into the container? -v /lib/firmware:/lib/firmware

@PatrickKutch Good catch. This is required if one wants to use E810 PF as the DPDK port. One VF this might not be needed. @dkosteck Can you  try without this mount when using the VF as the DPDK port on the testpmd server?

**Comment by dkosteck** [comment]
> client_cmd = [ "wget -O /tmp/client.py " "https://raw.githubusercontent.com/redhat-eets/netgauge/main/rfc2544/client.py" ] Will fail if we are doing tests in a air-gapped test environment (which we frequently do) - why is this not part of the container dockerfile?

Good point, I will change this

File: sriov/tests/config_template.yaml (hunk 3 above)
Comment (dkosteck) [reply sriov/tests/config_template.yaml]: Updated in latest

**Review by akiselev1 — COMMENTED** [review]
Can you update e2e.yaml CI flow to skip this test? And then run it to verify there is no regression elsewhere?

**Comment by dkosteck** [comment]
> Can you update e2e.yaml CI flow to skip this test? And then run it to verify there is no regression elsewhere?

@akiselev1 Updated e2e.yaml, will run it momentarily

**Pull request merged**